### ISSUE TITLE
How can I get more request context in WASM middleware?

### ISSUE THREAD

**Issue by chn-xu** [dapr/dapr#5003]
Hi, guys

I want to re-write the URL in WASM. For different scope service invocation. From [documents](https://docs.dapr.io/reference/components-reference/supported-middleware/middleware-wasm/) it seems like I can only get a request URL in the middleware. How can I get more request context in the middleware?​

**Comment by daixiang0** [comment]
Oh thanks for your interest, it is in plan, but implement is not hard. You can take a look and post a PR if you want.

**Comment by Taction** [comment]
I think it would be better to submit a proposal about the abi defination first, which defines the `Functions implemented in the Wasm module` and `Functions implemented in the host environment`. And maybe this can be implemented in another middleware, leaving the `basic` as a simple example for beginners. 
@daixiang0 WDYT?

@GeekEz also you can do this based on your requirements, if this is hard for you I can promote a draft idea and we can move forward based on that.

**Comment by mcandeia** [comment]
I agree with @Taction, how about [proxy-wasm](https://github.com/proxy-wasm/spec) ABI especification?

I don't know if it's too much for our middlewares but it supports a lot of operations and it is already a standard. Having yet the possibility to support envoy proxies out-of-the-box and being the second one to implement a server supporting proxy-wasm spec.

**Comment by Taction** [comment]
The [proxy-wasm](https://github.com/proxy-wasm/spec) ABI is definitely a good idea! And there is a [proxy-wasm-go-sdk](https://github.com/tetratelabs/proxy-wasm-go-sdk) provided by tetratelabs and a [proxy-wasm-go-host](https://github.com/mosn/proxy-wasm-go-host) provided by mosn. I've read the source code previously and adding wazero runtime support for the host is not hard, but there are still two things need to be taken into consideration:

1. The proxy-wasm-go-host may have not catch up with the latest proxy-wasm abi spec (But I haven't followed the project recently and I'm not quite sure now).
2. This may be a big work, and need someone knows or willing to learn about this, finish the code and doc and maybe a simple example.

Besides, a simpler abi may contains less code and easier to complete. Also can be implemented as another middleware. We don't need to choose only one about it, and these two middleware can be implemented in parallel, depending on if there is anyone interested in working on it.

**Comment by GeekEz** [comment]
> I think it would be better to submit a proposal about the abi defination first, which defines the `Functions implemented in the Wasm module` and `Functions implemented in the host environment`. And maybe this can be implemented in another middleware, leaving the `basic` as a simple example for beginners. @daixiang0 WDYT?
> 
> @GeekEz also you can do this based on your requirements, if this is hard for you I can promote a draft idea and we can move forward based on that.

Yes, please.

**Comment by daixiang0** [comment]
Actually, in the beginning, I name it "basic" because I want it to keep easy and simple, proxy-wasm is good but very complex and I want to implement it in another WASM middleware.

I want to get more feedback about just leaving "basic" for beginners as an example or still making it cover the whole request/response context.

/cc @dapr/maintainers-dapr @dapr/maintainers-components-contrib

**Comment by Taction** [comment]
Following is a simplified proposal for basic middleware. Another idea is to keep L7 abi the same as proxy-wasm, but since we can impl a proxy-wasm middleware, so I prefer to keep this simpler.
/cc @GeekEz @daixiang0 


# Functions implemented in the Wasm module

### `_start`

- params:
  - none
- returns:
  - none

Start function which is called when the module is loaded and initialized. This can be used by SDKs to setup and/or initialize state, but no proxy_ functions can be used at that point yet.

### `proxy_on_request`

- params:
  - `i32 (size_t) num_headers`
  - `i32 (size_t) body_size`
- returns:
  - `i32 (proxy_action_t) next_action`

Called when HTTP request is received from the client. Headers can be retrieved using `proxy_get_header` and/or `proxy_get_header_value`.Request body can be retrieved using `proxy_get_body`. The return vaule `next_action` currently have one valid value `ActionContinue` means that the host continues the processing.

### `proxy_on_response`

- params:
  - `i32 (size_t) num_headers`
  - `i32 (size_t) body_size`
- returns:
  - `i32 (proxy_action_t) next_action`

Called when HTTP response is received from the upstream. Headers can be retrieved using `proxy_get_header` and/or `proxy_get_header_value`. Response body can be retrieved using `proxy_get_body`.



# Functions implemented in the host environment

All functions implemented in the host environment return `proxy_result_t`, which indicates the status of the call (successful, invalid memory access, etc.), and the return values are written into memory pointers passed in as arguments (indicated by the `return_` prefix in the specification).

## Logging

### `proxy_log`

- params:
  - `i32 (proxy_log_level_t) log_level`
  - `i32 (const char*) message_data`
  - `i32 (size_t) message_size`
- returns:
  - `i32 (proxy_result_t) call_result`

Log message (`message_data`, `message_size`) at the given `log_level`.

## Requests and response

### `proxy_get_body`

- params:
  - `i32 (proxy_buffer_type_t) buffer_type`
  - `i32 (offset_t) offset`
  - `i32 (size_t) max_size`
  - `i32 (const char**) return_buffer_data`
  - `i32 (size_t*) return_buffer_size`
  - `i32 (uint32_t*) return_flags`
- returns:
  - `i32 (proxy_result_t) call_result`

Get up to max_size bytes from the `buffer_type`(enum for request and response), starting from `offset`. Bytes are written into buffer slice (`return_buffer_data`, `return_buffer_size`), and buffer flags are written into `return_flags`.

### `proxy_set_body`

- params:
  - `i32 (proxy_buffer_type_t) buffer_type`
  - `i32 (offset_t) offset`
  - `i32 (size_t) size`
  - `i32 (const char*) buffer_data`
  - `i32 (size_t) buffer_size`
  - `i32 (uint32_t) flags`
- returns:
  - `i32 (proxy_result_t) call_result`

Set content of the buffer `buffer_type` to the bytes (`buffer_data`, `buffer_size`), replacing `size` bytes, starting at `offset` in the existing buffer.

### `proxy_get_header`

- params:
  - `i32 (proxy_map_type_t) map_type`
  - `i32 (const char**) return_map_data`
  - `i32 (size_t*) return_map_size`
- returns:
  - `i32 (proxy_result_t) call_result`

Get all key-value pairs from a given map with `map_type`(enum for request and response).

### `proxy_set_header`

- params:
  - `i32 (proxy_map_type_t) map_type`
  - `i32 (const char*) map_data`
  - `i32 (size_t) map_size`
- returns:
  - `i32 (proxy_result_t) call_result`

Set all key-value pairs in a given map (`map_type`).

### `proxy_get_header_value`

- params:
  - `i32 (proxy_map_type_t) map_type`
  - `i32 (const char*) key_data`
  - `i32 (size_t) key_size`
  - `i32 (const char**) return_value_data`
  - `i32 (size_t*) return_value_size`
- returns:
  - `i32 (proxy_result_t) call_result`

Get content of key (`key_data`, `key_size`) from a given map (`map_type`).

### `proxy_set_header_value`

- params:
  - `i32 (proxy_map_type_t) map_type`
  - `i32 (const char*) key_data`
  - `i32 (size_t) key_size`
  - `i32 (const char*) value_data`
  - `i32 (size_t) value_size`
- returns:
  - `i32 (proxy_result_t) call_result`

Set or replace the content of key (`key_data`, `key_size`) to the value (`value_data`, `value_size`) in a given map (`map_type`).

### `proxy_add_header_value`

- params:
  - `i32 (proxy_map_type_t) map_type`
  - `i32 (const char*) key_data`
  - `i32 (size_t) key_size`
  - `i32 (const char*) value_data`
  - `i32 (size_t) value_size`
- returns:
  - `i32 (proxy_result_t) call_result`

Add key (`key_data`, `key_size`) with the value (`value_data`, `value_size`) to a given map (`map_type`).

### `proxy_remove_header_value`

- params:
  - `i32 (proxy_map_type_t) map_type`
  - `i32 (const char*) key_data`
  - `i32 (size_t) key_size`
- returns:
  - `i32 (proxy_result_t) call_result`

Remove key (`key_data`, `key_size`) from a given map (`map_type`).

### `proxy_send_http_response`

- params:
  - `i32 (uint32_t) response_code`
  - `i32 (const char*) response_code_details_data`
  - `i32 (size_t) response_code_details_size`
  - `i32 (const char*) response_body_data`
  - `i32 (size_t) response_body_size`
  - `i32 (const char*) additional_headers_map_data`
  - `i32 (size_t) additional_headers_size`
- returns:
  - `i32 (proxy_result_t) call_result`

Sends HTTP response without forwarding request to the upstream.

**Comment by daixiang0** [comment]
@Taction pretty cool, are you interested in implementing it?

**Comment by codefromthecrypt** [comment]
Thanks for looping me in. For starters, I think it would be easier to develop this as a fasthttp based project then after experimentation add it to dapr. The feedback loop would be a lot faster IMHO, especially as this seems like basically implementing fasthttp middleware in wasm. What do you think?

**Comment by anuraaga** [comment]
Heard about this issue from @codefromthecrypt - lately my main focus has been on WASM middleware for Envoy (proxy-wasm), WAF functionality via the [Coraza](https://github.com/corazawaf/coraza) project so can provide some perspective from that angle. Coraza is interested in being well integrated with many server projects as possible to deliver WAF functionality where users need it, but if every high level framework exposed their own middleware SDK or WASM ABI, it would end up an explosion of coraza middleware implementations. In some sense, implementing the filter with proxy-wasm and supporting any server implementing proxy-wasm would be really cool in principle by being very general. In practice, it's true that proxy-wasm was defined somewhat 1:1 with Envoy's processing model and is hard if not impossible to implement in other frameworks, here too, so it isn't actually very general :)

A middle-ground of middleware for a lower level framework such as fasthttp would be great if that could allow middleware such as WAF to be generally applicable within Go frameworks, where it is used somewhat commonly. Would be happy to help with the effort if that seems like a direction to take here.

**Comment by codefromthecrypt** [comment]
copying in @erikdubbelboer from fasthttp. I feel like this is a convergence of fans who would like to help get a webassembly plugin thing working with a primary consumer of at least dapr. The key here is folks not just wanting it to happen but offering to help. I would also help, and prefer to solve this first at the fasthttp abstraction, if the project is interested.

**Comment by codefromthecrypt** [comment]
fyi made a blank repo to play around with this as I think trying to design everything in one issue will be pretty crazy. @Taction feel free to move your initial strawman over if you like, you have admin https://github.com/http-wasm/http-wasm-abi

**Comment by codefromthecrypt** [comment]
ps we (tetrate) are going to setup infrastructure in this org towards whatever end. there are years of pressure building up to solve this, just no one did it. waiting isn't really all that helpful as it just reinforces go as the slower leg in webassembly progress.

the most important thing will be a -benchmark repository which can show some of the dire performance implications as well incidental complexity of whatever the work-in-progress model is vs proxy-wasm (in go). This way it is much less abstract and also easier to show the performance implications which if are not measured can easily result in ABI unusable for things like auth requests. anyone interested can follow the org and feel free to also not choose this path and continue down this issue another way!

**Comment by daixiang0** [comment]
Another thing needs to be considered -- now FastHTTP still does not support HTTP2 in production, as https://github.com/dapr/dapr/issues/4979 said.

For the first step, HTTP1 support is enough, but for the future, we need to consider more.

@erikdubbelboer what is status of HTTP2 support in FastHTTP?

**Comment by codefromthecrypt** [comment]
my 2p I think HTTP semantics are still defined in terms of the text protocol. When we start dropping down to the binary framing, it starts looking like a proxy again. I might focus on HTTP semantics at least for starters because that's typically the abstraction middleware are defined (not just TinyGo).

**Comment by Taction** [comment]
Cool! I agree with @codefromthecrypt and @anuraaga, we can implement that as a lower-level framework middleware(for now it is fasthttp, and can be implemented to other frameworks) and use it in dapr when finished. I will move my draft idea to the repo later.

As for http2, dapr is considering migrating to net/http, and middleware can be implemented for net/http that supports http2. And yes, HTTP1 support is enough for the first step.

**Comment by anuraaga** [comment]
Just a note for HTTP/2 vs 1, defining the ABI in terms of text should work fine for both since most frameworks do that anyways (e.g., headers are strings in almost all of them). We'll probably just need to define some semantics on how to present information that translates between them, e.g. possibly requiring implementations of the ABI to translate `Host` -> `:authority`. Not doing so could make it harder to write middleware that deals with headers. But such semantics would probably be steps down from the wasm ABI itself so just something to keep in mind.

**Comment by pkedy** [comment]
Does anyone think that having several fine-grained calls over the FFI boundary would be a performance issue? That is definitely my concern. I'm curious if proxy-wasm is this fine-grained. It might make sense to have the Wasm module return a control structure with modifications to make in the FastHTTP request and response?

I see that the basic Wasm middleware component uses [wapc](https://wapc.io) but only for redirects. It is capable of more. Check out this example for a general sense of what I'm talking about. I think it needs "more" but just wanted to get the general idea shown. https://github.com/wapc/examples/tree/main/middleware

**Comment by daixiang0** [comment]
@pkedy definitely it can do more than redirects! Before adding more functions, I want to discuss whether we need to introduce a new WASM ABI or reuse proxy-wasm ABI.

Performance is always in my mind but we need to trade-off between performance and flexibility in most cases.

**Comment by pkedy** [comment]
@daixiang0 Fair point. I'd like to understand that myself. Maybe others can chime on regarding proxy-wasm.

**Comment by pkedy** [comment]
@daixiang0 I've worked on HTTP1/HTTP2/gRPC proxies similar to Envoy. I remember now that Envoy filters work at that protocol/frame level (e.g. OnHeaders, OnBody, etc) as they are sent over the wire. I think the difference in this scenario is Dapr is operating on the HTTP context struct *after* it has been processed by fasthttp (or net/http). So I'm guessing that proxy-wasm might not be a fit based on the API surface of fasthttp. I believe @anuraaga made this the point above. I'm just saying it differently ;)

**Comment by daixiang0** [comment]
Yes, proxy-wasm has many coupled with Envoy itself, although reuse can introduce new ecological support, it also brings huge complexity.

Let's start by a new ABI and make it simple and easy!

**Comment by erikdubbelboer** [comment]
> @erikdubbelboer what is status of HTTP2 support in FastHTTP?

HTTP2 will probably never be supported by fasthttp. To be honest I would just switch to `net/http` as soon as possible for this project.

**Comment by anuraaga** [comment]
> Does anyone think that having several fine-grained calls over the FFI boundary would be a performance issue? That is definitely my concern. I'm curious if proxy-wasm is this fine-grained.

`proxy-wasm` is fine-grained in the sense that host calls are used to get data about the request such as headers, body. FFI is a bit slower than a normal function call, but not a whole lot more - given the wide range of middleware, and even the wide range of languages (e.g. Rust vs TinyGo) with different performance characteristics, it's hard to make any definite recommendation, but I think generally having a few more FFI to lazily access data from the host won't be significant cost and in most middleware is less overall work. It would take a whole lot of them to outweigh having to serialize deserialize the request / response across boundaries whether the serialization is HTTP, JSON, msgpack, or anything it's still quite a lot of extra work. We'll just need to benchmark to verify though.

Just some points to keep in mind

- Many middleware by definition won't need much of the request. Auth middleware almost always just reads a single header, rate limiting also either nothing or just URI and maybe another header. Always need benchmark to verify but it's unlikely that a fine-grained approach would be slower than copying the entire request.
- Proxy middleware will just want to forward HTTP within the host. Using FFI with handles to host objects will be simpler to implement this without copying data into wasm without any processing.
- If going with an approach that copies the entire request into the middleware, it will likely be worth having a structured ABI rather than a single string, e.g. instead of parsing out msgpack, it would be simpler to read method, uri, headers from separate arguments in the exported function than having a single string

Let's try things out and see how it goes. Glad to see ideas for a new simpler HTTP ABI!

**Comment by pkedy** [comment]
@anuraaga I think my example created more confusion than clarity ;) Forget about waPC and MsgPack. My general point is that like how `proxy_set_headers` lets you set multiple headers in a single FFI call, I think it would be simpler _and_ faster to collapse at least all header mutations into a single (and very simple) binary payload representation. Something like this seems already necessary for encoding `map_data`. Right?

From that perspective, what I'm proposing is a minor tweak. `proxy_set_header`, `proxy_set_header_value`, `proxy_add_header_value`, `proxy_remove_header_value` and possibly `proxy_send_http_response` would be replaced with a single FFI call. Inside the middleware, there could be more fine-grained set/add/remove calls that keep the edits in memory until its time to cross the FFI boundary.

> Many middleware by definition won't need much of the request... Always need benchmark to verify but it's unlikely that a fine-grained approach would be slower than copying the entire request.

100% agree. Most only need a header or two. My thought here was that if the host knows which headers the middleware needs, it can be sent in them during the initial call and potentially avoid further FFI calls. This makes some assumptions though and does add complexity.

Performance is important to think about in the context of HTTP middleware. For better or worse, developers tend to be (overly?) concerned about it in their proxies and http libraries. I found that the approach of batching things into simple control structures greatly improves performance and that could be helpful in this scenario also. To your point, I think would be interesting and valuable to benchmark this to understand the tradeoffs (e.g. simplicity and performance).

**Comment by pkedy** [comment]
Just as a reference: On my M1 Mac, I benched a very simple call from host to an empty guest function at ~260ns, which is insanely good. Kudos to the wazero team! This may not be as much of a concern as it once was.

**Comment by daixiang0** [comment]
Now @ItalyPaleAle is working on converting Dapr middlewares and HTTP channel from fasthttp to net/http, we do not need to consider fasthttp now.

**Comment by codefromthecrypt** [comment]
thanks for the ping. In the finalizing proof-of-concept, we'll focus on net/http though having fasthttp is still useful for modeling. Also the interaction of the fasthttp adaptor with net/http is interesting as the response writing is deferred. good context for design thinking.

**Comment by ItalyPaleAle** [comment]
@codefromthecrypt I would recommend ignoring fasthttp as we'll eventually migrate off that for a variety of reasons (see #4979). In particular with regards to the response writing, fasthttp is doing it wrong, as it doesn't allow sending data to the client as a stream: it requires all data to be saved in memory before it begins flushing it out.

**Comment by codefromthecrypt** [comment]
@ItalyPaleAle ack yeah that buffering part is part of the issue, as if anything it should be opt-in.

For the case of dapr I'll ignore fasthttp and buffering, even if some wasm plugins need to buffer (or at least to be able to attempt to buffer).

**Comment by codefromthecrypt** [comment]
in case curious (not immediate need here), Web Application Firewall (WAF) can contain rules that require buffering, and so for example its wasm has configuration around that. OTOH supporting this is more about what the middleware is capable of doing vs whether or not fasthttp is involved. https://github.com/corazawaf/coraza-proxy-wasm

**Comment by codefromthecrypt** [comment]
@chn-xu can you take a look at the examples of the work in progress tinygo SDK for full access to HTTP?

https://github.com/http-wasm/http-wasm-guest-tinygo

We have a working implementation integrated with dapr (and mosn also), but want to make sure this has the functions needed before we wrap up for a PR

https://github.com/dapr/components-contrib/compare/master...http-wasm:components-contrib:httpwasm

**Comment by codefromthecrypt** [comment]
raised this as noticed we are missing a feature for logging https://github.com/dapr/kit/pull/24

**Comment by chn-xu** [comment]
> @chn-xu can you take a look at the examples of the work in progress tinygo SDK for full access to HTTP?
> 
> https://github.com/http-wasm/http-wasm-guest-tinygo
> 
> We have a working implementation integrated with dapr (and mosn also), but want to make sure this has the functions needed before we wrap up for a PR
> 
> [dapr/components-contrib@master...http-wasm:components-contrib:httpwasm](https://github.com/dapr/components-contrib/compare/master...http-wasm:components-contrib:httpwasm)

Yeah, it looks great.

**Comment by codefromthecrypt** [comment]
cool. documentation is here now, though I owe some diagrams (will make a site for it)
https://github.com/http-wasm/http-wasm-abi/blob/main/http-handler/http-handler.md

I'll add a PR in the next few days to contrib

**Comment by codefromthecrypt** [comment]
We've significantly improved the speed of wazero, so will raise the PR once the new pre-release out is in (in a couple days). Note: this also speeds up wapc. For example, the current wapc basic URL rewrite benchmarks just under 2.8us/op where native is 1.0us/op. The same in the new ABI is 2.2us/op, or as fast as 1.7us if you use wasm directly instead of tinygo.  That said, the point of the new host is flexibility, not speed. Speed just happens to be better now due to wazero improvements.

**Comment by codefromthecrypt** [comment]
here's the PR, which is awaiting another to be merged first https://github.com/dapr/components-contrib/pull/2239 

While I raised this PR, @anuraaga deserves a great deal of credit having supported this, troubleshot and co-designed some parts of it as well.

**Comment by codefromthecrypt** [comment]
here's a WIP sample, which is mainly awaiting folks to decide it if is appropriate or not https://github.com/dapr/samples/pull/122
